File: client-jss/src/app/main/services/tag.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { AppRestService } from '../../services/appRest.service';
import { Author } from '../model/Author';
import { Category } from '../model/Category';
import { JssCategory } from '../model/JssCategory';
import { PublishingDepartment } from '../model/PublishingDepartment';
import { Serie } from '../model/Serie';
import { Tag } from '../model/Tag';

@Injectable({
  providedIn: 'root'
})
export class TagService extends AppRestService<Tag> {

  constructor(http: HttpClient) {
    super(http, "wordpress");
  }

  getAvailableTags() {
    return this.getListCached(new HttpParams(), "tags");
  }

  getTagBySlug(slug: string) {
    return this.get(new HttpParams().set("slug", slug), "tag/slug");
  }

  getAllTagsByJssCategory(jssCategory: JssCategory, isDisplayNewPosts: boolean) {
    return this.getList(new HttpParams().set("jssCategoryId", jssCategory.id).set("isDisplayNewPosts", isDisplayNewPosts), "tags/all/jss-category");
  }

  getAllTagsByCategory(page: number, size: number, category: Category) {
    return this.getPagedList(new HttpParams().set("page", page).set("size", size).set("categoryId", category.id), "tags/all/category");
  }

  getAllTagsByTag(tag: Tag, isDisplayNewPosts: boolean) {
    return this.getList(new HttpParams().set("tagSlug", tag.slug).set("isDisplayNewPosts", isDisplayNewPosts), "tags/all/tag");
  }

  getAllTagsByAuthor(author: Author, isDisplayNewPosts: boolean) {
    return this.getList(new HttpParams().set("authorSlug", author.slug).set("isDisplayNewPosts", isDisplayNewPosts), "tags/all/author");
  }

  getAllTagsByPremiumPosts() {
    return this.getList(new HttpParams(), "tags/all/premium");
  }

  getAllTagsBySerie(serie: Serie) {
    return this.getList(new HttpParams().set("serieSlug", serie.slug), "tags/all/serie");
  }

  getAllTagsByPublishingDepartment(department: PublishingDepartment) {
    return this.getList(new HttpParams().set("departmentCode", department.code), "tags/all/publishing-department");
  }

  getAllTendencyTags() {
    return this.getList(new HttpParams(), "tags/tendency");
  }

  getAllLastPostsTags() {
    return this.getList(new HttpParams(), "tags/last");
  }

  getAllMostSeenPostsTags() {
    return this.getList(new HttpParams(), "tags/most-seen");
  }
}

File: src/main/java/com/jss/osiris/modules/myjss/wordpress/service/TagService.java
package com.jss.osiris.modules.myjss.wordpress.service;

import java.time.LocalDateTime;
import java.util.List;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import com.jss.osiris.libs.exception.OsirisException;
import com.jss.osiris.modules.myjss.wordpress.model.Author;
import com.jss.osiris.modules.myjss.wordpress.model.Category;
import com.jss.osiris.modules.myjss.wordpress.model.JssCategory;
import com.jss.osiris.modules.myjss.wordpress.model.PublishingDepartment;
import com.jss.osiris.modules.myjss.wordpress.model.Serie;
import com.jss.osiris.modules.myjss.wordpress.model.Tag;

public interface TagService {
        public Tag getTag(Integer id);

        public Tag addOrUpdateTagFromWordpress(Tag tag);

        public List<Tag> getAvailableTags();

        public Tag getTagBySlug(String slug);

        public List<Tag> getAllTagsByJssCategory(JssCategory jssCategory);

        public List<Tag> getAllTagsByTag(Tag tag, LocalDateTime consultationDate) throws OsirisException;

        public Page<Tag> getAllTagsByCategory(Pageable pageable, Category category);

        public List<Tag> getAllTagsByAuthor(Author author, LocalDateTime consultationDate);

        public List<Tag> getAllTagsBySerie(Serie serie);

        public List<Tag> getAllTagsByPublishingDepartment(PublishingDepartment publishingDepartment)
                        throws OsirisException;

        public List<Tag> getAllTagsByIdf() throws OsirisException;

        public List<Tag> getAllTendencyTags() throws OsirisException;

        public List<Tag> getFollowedTagsForCurrentUser();
}
